### PR TITLE
Dockerfile: add support for building for ARM64 architecture.

### LINKED ISSUE #9898: Cilium enable on aarch64
We have ported Cilium master version on Arm server and deployed Nginx to verify and test k8s cluster that composed by three Arm nodes, the status of the pods and the connection between the nodes are normal.
I hope to contribute to cilium support for Arm64, and Related work includes the following aspects:
1.	Fix compiling and running issue on Arm64;
2.	Add arm64 support for building cilium-related images;
3.	Add multi-arch support for cilium-related images;
4.	Add arm64 support for cilium CI/CD.

### PR DESCRIPTION
**Pull request by nickb937**
Dockerfiles do not have good support for architecture based decisions
so case statements need to be used.

The changes:
- no libc6-dev-i386 on ARM
- cilium/bpf-map is built from source rather than downloading the binary
  image
- add more RUN statements so the whole package does not have to be
  rebuilt when fixing individual build stages.

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](http://docs.cilium.io/en/stable/contributing/contributing/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue. 
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](http://docs.cilium.io/en/stable/contributing/contributing/#dev-coo).
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #9898 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9969)

### CODE REVIEW

**Comment by mrostecki** [comment]
Not having `libc6-dev-i386` package installed usually resulted in the error mentioned here: #368. Another related issue: #8529

That seems like a bug in glibc for me. `gnu/stubs.h´ includes `gnu/stubs-32.h` even on non-x86_32 architectures. I think that macros of ifdefs related to architectures need to be fixed there. The tricky point is that `gnu/stubs.h` has no content initially:

https://sourceware.org/git/?p=glibc.git;a=blob;f=include/gnu/stubs.h;h=6308e4e0e04753a8d4c92074e3661fa17a0293f1;hb=HEAD

but it gets generated during build and in the most of distros it looks like:

```
#if !defined __x86_64__
# include <gnu/stubs-32.h>
#endif
#if defined __x86_64__ && defined __LP64__
# include <gnu/stubs-64.h>
#endif
#if defined __x86_64__ && defined __ILP32__
# include <gnu/stubs-x32.h>
#endif
```

Which looks obviously wrong, because aarch64 or ppc are going to satisfy the first `if` and include `gnu/stubs-32.h`.

@nickb937 If you are interested in debugging this, we can brainstorm next week. If not, I will try to look at the issue together with ARM people in openSUSE community. We are trying to get aarch64 builds for Cilium too, although I have also an another blocker - Envoy fails to build with aarch64 on our infra.

**Review by vadorovsky — COMMENTED** [review]
Not having `libc6-dev-i386` package installed usually resulted in the error mentioned here: #368. Another related issue: #8529

That seems like a bug in glibc for me. `gnu/stubs.h` includes `gnu/stubs-32.h` even on non-x86_32 architectures. I think that macros or ifdefs related to architectures need to be fixed there. The tricky point is that `gnu/stubs.h` has no content initially:

https://sourceware.org/git/?p=glibc.git;a=blob;f=include/gnu/stubs.h;h=6308e4e0e04753a8d4c92074e3661fa17a0293f1;hb=HEAD

but it gets generated during build and in the most of distros it looks like:

```
#if !defined __x86_64__
# include <gnu/stubs-32.h>
#endif
#if defined __x86_64__ && defined __LP64__
# include <gnu/stubs-64.h>
#endif
#if defined __x86_64__ && defined __ILP32__
# include <gnu/stubs-x32.h>
#endif
```

Which looks obviously wrong, because aarch64 or ppc are going to satisfy the first `if` and include `gnu/stubs-32.h`.

@nickb937 If you are interested in debugging this, we can brainstorm next week. If not, I will try to look at the issue together with ARM people in openSUSE community. We are trying to get aarch64 builds for Cilium too, although I have also an another blocker - Envoy fails to build on aarch64 on our infra.

**Comment by nickb937** [comment]
I have only been trying to build through the docker build method and so far have a comple package built successfully, but untested, for ARM64 (I have two further pull requests to submit). I have not experienced the stubs issue you refer to, but will look at my build to see why that isn't occurring.

I am interested in getting Cilium and Istio working to create a Kubernetes cluster on some Raspberry Pi4s, so I have a desire to make progress with these small issues 
I have a working blaze binary and made some progress with Envoy before switching my efforts to getting Cilium working first.

**Comment by nickb937** [comment]
The Docker build is sourced from Ubuntu 18.04.  Looking in the container FS, there is only `aarch64-linux-gnu/gnu/stubs.h` which contains:

```
#include <bits/wordsize.h>

#if !defined __AARCH64EB__
# include <gnu/stubs-lp64.h>
#endif
#if defined __AARCH64EB__
# include <gnu/stubs-lp64_be.h>
#endif
```

The difference I see between X86 and AARCH64 is that /usr/include/gnu does not exist on the latter, instead it is under /usr/include/aarch64-linux-gnu.  The compiler search paths are similarly different to reflect this:

AARCH64:
```
$ echo | gcc -E -Wp,-v -
ignoring nonexistent directory "/usr/local/include/aarch64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/aarch64-linux-gnu/7/../../../../aarch64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/aarch64-linux-gnu/7/include
 /usr/local/include
 /usr/lib/gcc/aarch64-linux-gnu/7/include-fixed
 /usr/include/aarch64-linux-gnu
 /usr/include

```

X86::

```
 echo | gcc -E -Wp,-v -
ignoring nonexistent directory "/usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../x86_64-pc-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include
 /usr/local/include
 /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include-fixed
 /usr/include
End of search list.
```

**Comment by mrostecki** [comment]
@nickb937 OK, then `gnu/stubs.h` looks good on Ubuntu (**except** x86_64, because it keeps including the 32bit header, a problem which we "solve" by installing the 32bit devel package).

All good then, your PR should work.

File: contrib/packaging/docker/Dockerfile.runtime [hunk 1]
```diff
@@ -88,6 +97,12 @@ cd /go/src/github.com/google/gops && \
 git checkout -b v0.3.6 v0.3.6 && \
 go install && \
 strip /go/bin/gops
+#
+# bpf-map
+#
+RUN go get -d github.com/cilium/bpf-map && \
+cd /go/src/github.com/cilium/bpf-map && \
+go build && strip bpf-map
```
Comment (aanm) [inline contrib/packaging/docker/Dockerfile.runtime]: Why this addition? This is pointing to master and not to a specific version like it was set on lines 66-68 from the previous version of this file.

**Review by aanm — CHANGES_REQUESTED** [review]

File: Dockerfile.builder [hunk 2]
```diff
@@ -19,7 +19,14 @@ ENV GO_VERSION 1.13.6
...
 # Build dependencies
 #
-RUN apt-get update \
+RUN dpkgArch="$(dpkg --print-architecture)"; \
+case "${dpkgArch##*-}" in \
+  arm64) libcdev="" ;; \
+  amd64) libcdev="libc6-dev-i386" ;; \
+  i386) libcdev="libc6-dev-i386" ;; \
+  *) libcdev="" ;; \
+esac; \
```
Comment (tklauser) [inline Dockerfile.builder:28]: Installing libc6-dev-386 should no longer be needed after #10204 being merged. We no longer depend on headers from libc.

File: contrib/packaging/docker/Dockerfile.runtime [hunk 3]
```diff
@@ -7,8 +7,15 @@ apt-get upgrade -y && \
 #
 # Prepackaged Cilium runtime dependencies
 #
+dpkgArch="$(dpkg --print-architecture)"; \
+case "${dpkgArch##*-}" in \
+  arm64) libcdev="" ;; \
+  amd64) libcdev="libc6-dev-i386" ;; \
+  i386) libcdev="libc6-dev-i386" ;; \
+  *) libcdev="" ;; \
+esac; \
```
Comment (tklauser) [inline contrib/packaging/docker/Dockerfile.runtime:16]: Same here, libc-dev-i386 is no longer needed after #10204

File: contrib/packaging/docker/Dockerfile.runtime [hunk 4]
```diff
@@ -41,35 +48,37 @@ apt-get install -y --no-install-recommends make git curl ca-certificates xz-util
 # Additional iproute2 build dependencies
   gcc git pkg-config bison flex build-essential \
 # Additional bpftool dependencies
-  python3 && \
+  python3
 #
 # iproute2
 #
+RUN \
```
Comment (tklauser) [inline contrib/packaging/docker/Dockerfile.runtime]: What's the reason for this change? Each `RUN` instruction will add a new layer and increase image size.

File: contrib/packaging/docker/Dockerfile.runtime [hunk 5]
```diff
@@ -41,35 +48,37 @@ apt-get install -y --no-install-recommends make git curl ca-certificates xz-util
...
 ./configure && \
 make -j `getconf _NPROCESSORS_ONLN` && \
 strip tc/tc && \
-strip ip/ip && \
-cd .. && \
+strip ip/ip
 #
 # bpftool
 #
+RUN \
```
Comment (tklauser) [inline contrib/packaging/docker/Dockerfile.runtime]: Same here. Why introduce a separate layer for this?

File: contrib/packaging/docker/Dockerfile.runtime [hunk 6]
```diff
@@ -41,35 +48,37 @@ apt-get install -y --no-install-recommends make git curl ca-certificates xz-util
...
-#
-curl -SsL https://github.com/cilium/bpf-map/releases/download/v1.0/bpf-map -o bpf-map && \
-chmod +x bpf-map && \
-strip bpf-map && \
+strip bpftool
 #
 # cni/loopback
 #
-curl -sS -L https://github.com/containernetworking/plugins/releases/download/v0.7.5/cni-plugins-amd64-v0.7.5.tgz -o cni.tar.gz && \
+RUN \
```
Comment (tklauser) [inline contrib/packaging/docker/Dockerfile.runtime]: And again here.

File: contrib/packaging/docker/Dockerfile.runtime (hunk 1 above)
Comment (tklauser) [reply contrib/packaging/docker/Dockerfile.runtime]: It looks like this was moved from above. @nickb937 what was the reason to move this here? Also this adds another layer due to the additional `RUN`.

**Review by tklauser — CHANGES_REQUESTED** [review]
Please rebase this PR on top of the latest master to get the libc changes from #10204

File: contrib/packaging/docker/Dockerfile.runtime (hunk 4 above)
Comment (aanm) [reply contrib/packaging/docker/Dockerfile.runtime]: it's in the PR description:
>add more RUN statements so the whole package does not have to be
rebuilt when fixing individual build stages

Comment (tklauser) [reply contrib/packaging/docker/Dockerfile.runtime]: Thanks for the pointer, somehow overlooked this. So then it's mainly a trade-off between incremental build speed and image size. Do we in practice need to rebuild often while fixing individual build stages (outside of this PR)?

**Comment by iecedge** [comment]
@nickb937  I can help update and rebase this patch if you don't have time.

**Comment by joestringer** [comment]
@nickb937 @iecedge do you think that we'll be able to respin this PR soon?

File: contrib/packaging/docker/Dockerfile.runtime (hunk 1 above)
Comment (nickb937) [reply contrib/packaging/docker/Dockerfile.runtime]: IIRC, the bpf-map was moved here because it needed `go` to build from sources, which wasn't available in its previous location.

**Comment by nickb937** [comment]
I'm currently testing a rebase.  It will take 1-2 days to complete the build

**Comment by joestringer** [comment]
@nickb937 @Jianlin-lv can you take a look at this PR and see if there is anything left in this PR that improves the version that's merged on master?

**Comment by Jianlin-lv** [comment]
> @nickb937 @Jianlin-lv can you take a look at this PR and see if there is anything left in this PR that improves the version that's merged on master?

I think the content mentioned in this PR is already supported on the master.

**Comment by joestringer** [comment]
Thanks for the assessment @Jianlin-lv . And thanks again @nickb937 for proposing this PR, I think it helped a lot to identify pain points in the build system for ARM64 support. If you see any further changes that could help Cilium on ARM64 then feel free to file an issue or propose additional PRs.

Closing, superseded.
